### PR TITLE
Homework3

### PR DESCRIPTION
**Pull request by jelenanemcic**
To test the application:
- Initialize databases
- Run controller and calculator
- Run betgenerator script
- Run eventsettler script

I added log messages in controller and in calculator when we update the database according to the new event update, so that we can see that the database gets updated correctly.
I also commented out a few lines in controller that I thought were not needed.

### CODE REVIEW

File: lecture_3/03_project/controller/internal/engine/engine.go [hunk 1]
```diff
@@ -57,8 +57,8 @@ func (e *Engine) processBetsCalculated(ctx context.Context) error {
 		return err
 	}
 
-	resultingBets := e.handler.HandleBetsCalculated(ctx, consumedBetsCalculated)
-	e.publisher.PublishBets(ctx, resultingBets)
+	e.handler.HandleBetsCalculated(ctx, consumedBetsCalculated)
+	//	e.publisher.PublishBets(ctx, resultingBets)
```
Comment (oreskdaxilis) [inline lecture_3/03_project/controller/internal/engine/engine.go]: I suppose you have made this change in order not to produce the bet back to Calculator service since it already has the latest state of the bet. What you're missing here is checking whether the bet has changed or not, and not producing only if the bet is unchanged after handling the consumed message.
Why? Because `bets` queue is not a "queue used for communication between Controller and Calculator services"; it's rather a "queue where Controller is publishing the latest state of every bet" and every interested service can subscribe to the queue. It is just the case in our system that only one service is listening to those changes, but it could happen (and in our production betting platform it is the case) that there are several consumers of the same queue.
When you look at it that way, if the bet has changed in any way, there may be some listener interested in receiving the latest state of the bet, so you should produce it into the queue. It is the responsibility of the consumer to ignore the message not relevant for them (but based on what I see, you already have handled that correctly on the Calculator side).

File: lecture_3/03_project/calculator/internal/domain/models/betCalculated.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package models
+
+// BetCalculated represents a domain model representation of a calculated bet.
+type BetCalculated struct {
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/domain/models/betCalculated.go]: `Calculated` suffix is not needed here - from the perspective of this (micro)service, those are the "bets". Of course, data model for representing those bets is not identical to the one used in Controller, but that is not relevant here. Try to think about them as completely separate services which don't know anything about each other. They only know what kind of data they consume from RabbitMQ queues and what kind of data are they supposed to produce into the queue.
Calculator service is listening to `bets` queue where bets are incoming represented as `rabbitmqmodels.Bet` and this service will map it into domain `Bet` representation (which is this file), containing only the relevant subset of information.

That's also why the table in `calc_bets` database is actually called `bets`.

File: lecture_3/03_project/calculator/internal/engine/handler/betRepository.go [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+	"context"
+
+	domainmodels "github.com/superbet-group/code-cadets-2021/lecture_3/03_project/calculator/internal/domain/models"
+)
+
+type BetRepository interface {
+	InsertBetCalculated(ctx context.Context, bet domainmodels.BetCalculated) error
+	UpdateBetCalculated(ctx context.Context, bet domainmodels.BetCalculated) error
+	GetBetCalculatedByID(ctx context.Context, id string) (domainmodels.BetCalculated, bool, error)
+	GetBetCalculatedBySelectionID(ctx context.Context, id string) ([]domainmodels.BetCalculated, bool, error)
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/engine/handler/betRepository.go]: Once you rename `BetCalculated` struct as suggested in another comment, these methods should also be renamed accordingly.

Comment (oreskdaxilis) [reply lecture_3/03_project/calculator/internal/engine/handler/betRepository.go]: Also regarding `GetBetCalculatedBySelectionID`:
- there is no need to return `exists` flag, if there are no results you can just return an empty slice. Note that we cannot do the same with methods where only one value is being returned, because some kind of struct is guaranteed to be returned and you cannot tell whether the empty struct as a result was returned because it didn't exist in the repository or that's the actual value which was fetched.
- consider renaming `id string` to `selectionId string`; it sounds a bit better to me, even though your naming is also perfectly fine so feel free to keep it like this if you disagree.

File: lecture_3/03_project/calculator/internal/engine/handler/handler.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+import (
+	"context"
+	"log"
+
+	domainmodels "github.com/superbet-group/code-cadets-2021/lecture_3/03_project/calculator/internal/domain/models"
+	rabbitmqmodels "github.com/superbet-group/code-cadets-2021/lecture_3/03_project/calculator/internal/infrastructure/rabbitmq/models"
+)
+
+// Handler handles bets received and bets calculated.
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/engine/handler/handler.go]: Comment from Controller

File: lecture_3/03_project/calculator/internal/engine/handler/handler.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+// New creates and returns a new Handler.
+func New(betRepository BetRepository) *Handler {
+	return &Handler{
+		betRepository: betRepository,
+	}
+}
+
+// HandleBets handles bets received.
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/engine/handler/handler.go]: Comment from Controller

File: lecture_3/03_project/calculator/internal/engine/handler/handler.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+			// Calculate the calculated bet based on the incoming bet.
+			domainBet := domainmodels.BetCalculated{
+				Id:                   bet.Id,
+				SelectionId:          bet.SelectionId,
+				SelectionCoefficient: bet.SelectionCoefficient,
+				Payment:              bet.Payment * bet.SelectionCoefficient,
+			}
+
+			// Insert the calculated bet into the repository.
+			err := h.betRepository.InsertBetCalculated(ctx, domainBet)
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/engine/handler/handler.go]: Related to my comment about how `bets` queue is a queue where Controller publishes latest state of every bet, it is not sufficient here to "blindly" insert the bet into the repository - Calculator shouldn't guess if this is the first occurrence of the bet in `bets` queue or not. I suggest firstly trying to fetch the bet from the repository and then having a separate logic of handling bets which already existed in the repository and those which didn't.

File: lecture_3/03_project/calculator/internal/engine/handler/handler.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+				var resultingBet rabbitmqmodels.BetCalculated
+
+				// Calculate the resulting calculated bet, which should be published.
+				if eventUpdate.Outcome == "won" {
+					resultingBet = rabbitmqmodels.BetCalculated{
+						Id:     domainBet.Id,
+						Status: eventUpdate.Outcome,
+						Payout: domainBet.Payment,
+					}
+				} else {
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/engine/handler/handler.go]: Maybe it would be a bit better to have `else if eventUpdate.Outcome == "lost"` and then an `else` block where an error would be logged, but this is also fine.

File: lecture_3/03_project/calculator/internal/infrastructure/sqlite/models/betCalculated.go [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package models
+
+// BetCalculated is a storage model representation of a calculated bet.
+type BetCalculated struct {
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/infrastructure/sqlite/models/betCalculated.go]: Similar to my comment for domain `BetCalculated`, this should be renamed to `Bet`.

File: lecture_3/03_project/calculator/internal/infrastructure/sqlite/betRepository.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+	}
+}
+
+// InsertBetCalculated inserts the provided calculated bet into the database. An error is returned if the operation
+// has failed.
+func (r *BetRepository) InsertBetCalculated(ctx context.Context, bet domainmodels.BetCalculated) error {
+	storageBet := r.betMapper.MapDomainBetToStorageBet(bet)
+	err := r.queryInsertBet(ctx, storageBet)
+	if err != nil {
+		return errors.Wrap(err, "calc_bets repository failed to insert a bet with id "+bet.Id)
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/infrastructure/sqlite/betRepository.go]: It's up to you what you want to put in the error message, but I just want to point out I've used `"bet repository failed to..." `text in Controller because this struct is called `BetRepository`, not because it's operating on `bets` database. We usually use that convention in our production code, it is helpful for debugging to have the struct name within the error message.

File: lecture_3/03_project/calculator/internal/infrastructure/sqlite/betRepository.go [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+	}
+	return nil
+}
+
+func (r *BetRepository) queryInsertBet(ctx context.Context, bet storagemodels.BetCalculated) error {
+	// If payout is 0, do not insert it.
+	if bet.Payment == 0 {
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/infrastructure/sqlite/betRepository.go]: This is not needed. In Controller I wanted to check whether `payout` is 0 and not insert it into DB if it is. However, for `payment` we can assume it will never be 0 (because a bet with payment of 0 shouldn't even enter the system, that bet should be refused). You can safely insert in into the DB without having this check.

File: lecture_3/03_project/calculator/internal/infrastructure/sqlite/betRepository.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+func (r *BetRepository) queryUpdateBet(ctx context.Context, bet storagemodels.BetCalculated) error {
+	updateBetSQL := "UPDATE bets SET selection_id=?, selection_coefficient=?, payment=? WHERE id=?"
+
+	statement, err := r.dbExecutor.PrepareContext(ctx, updateBetSQL)
+	if err != nil {
+		return err
+	}
+
+	_, err = statement.ExecContext(ctx, bet.Payment, bet.Id)
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/infrastructure/sqlite/betRepository.go]: Is this going to work? You have only 2 arguments here and 4 arguments few lines above when creating `updateBetSQL`. Let me know if I'm missing something.

File: lecture_3/03_project/calculator/internal/infrastructure/sqlite/betRepository.go [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+		return storagemodels.BetCalculated{}, err
+	}
+	defer row.Close()
+
+	// This will move to the "next" result (which is the only result, because a single bet is fetched).
+	row.Next()
+
+	var selectionId string
+	var selectionCoefficient int
+	var paymentSql sql.NullInt64
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/infrastructure/sqlite/betRepository.go]: Payment will always be set, it is not nullable in the DB (check data model in README or in `dbinitializer` script).

File: lecture_3/03_project/calculator/internal/engine/publisher/betPublisher.go [hunk 13]
```diff
@@ -0,0 +1,11 @@
+package publisher
+
+import (
+	"context"
+
+	rabbitmqmodels "github.com/superbet-group/code-cadets-2021/lecture_3/03_project/calculator/internal/infrastructure/rabbitmq/models"
+)
+
+type BetPublisher interface {
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/engine/publisher/betPublisher.go]: Consider renaming this to `BetCalculatedPublisher` as it is publishing `BetCalculated` messages.

Note I've suggested renaming `domainmodels.BetCalculated` to `Bet` and `storagemodels.BetCalculated` to `Bet`, but in this case `BetCalculated` is the proper naming because it is a data model which should be produced to `bets-calculated` queue.

**Review by oreskdaxilis — COMMENTED** [review]
This solution is actually really good, I just didn't leave comments where you did everything correctly. 👍

File: lecture_3/03_project/calculator/internal/infrastructure/sqlite/betRepository.go (hunk 11 above)
Comment (jelenanemcic) [reply lecture_3/03_project/calculator/internal/infrastructure/sqlite/betRepository.go]: Thank you, I did not notice, this was a mistake. But the code worked because I actually never called UpdateBet function, now I am going to fix that as well.

Comment (oreskdaxilis) [reply lecture_3/03_project/calculator/internal/infrastructure/sqlite/betRepository.go]: Yep I know, no worries!

File: lecture_3/03_project/controller/internal/engine/engine.go (hunk 1 above)
Comment (jelenanemcic) [reply lecture_3/03_project/controller/internal/engine/engine.go]: I think I fixed all mistakes, I updated Controller so it first checks whether a bet was changed and if so, publishes it to bets queue. Also, Calculator now checks if there is already a bet with that ID in the database and if not, it inserts a new one, and if it does exist, Calculator updates it (I don't know if this is necessary, I think the bet should always be the same here). And when an event update is received, Calculator updates the calc_bets database (I know we said in the class that it is not necessary, but just to check that update function works) :)

File: lecture_3/03_project/calculator/internal/engine/handler/betRepository.go [hunk 14]
```diff
@@ -7,8 +7,8 @@ import (
...
 
 type BetRepository interface {
-	InsertBetCalculated(ctx context.Context, bet domainmodels.BetCalculated) error
-	UpdateBetCalculated(ctx context.Context, bet domainmodels.BetCalculated) error
-	GetBetCalculatedByID(ctx context.Context, id string) (domainmodels.BetCalculated, bool, error)
-	GetBetCalculatedBySelectionID(ctx context.Context, id string) ([]domainmodels.BetCalculated, bool, error)
+	InsertBet(ctx context.Context, bet domainmodels.Bet) error
+	UpdateBet(ctx context.Context, bet domainmodels.Bet) error
+	GetBetByID(ctx context.Context, id string) (domainmodels.Bet, bool, error)
+	GetBetBySelectionID(ctx context.Context, selectionId string) ([]domainmodels.Bet, error)
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/engine/handler/betRepository.go]: Sorry, I didn't notice this in the first review; this method should actually be called `GetBetsBySelectionID`, because you are fetching multiple bets.

File: lecture_3/03_project/calculator/internal/engine/handler/handler.go [hunk 15]
```diff
@@ -66,33 +88,35 @@ func (h *Handler) HandleEventUpdates(
...
 					}
-				} else {
+				} else if eventUpdate.Outcome == "lost" {
 					resultingBet = rabbitmqmodels.BetCalculated{
 						Id:     domainBet.Id,
 						Status: eventUpdate.Outcome,
 						Payout: 0,
 					}
+				} else {
+					log.Fatalf("Wrongly defined outcome: %s", eventUpdate.Outcome)
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/engine/handler/handler.go]: Don't call `Fatalf`, because that will crash the application, which is certainly not something handler should be supposed to do. The best effort from handler's side is to return an error and let the caller decide what to do with it. Since the problem occurred while processing a message from the input channel, you cannot return that error as a regular return value. What you could do is change the method signature to return `(<-chan rabbitmqmodels.BetCalculated, <-chan error)` and then you can write into the errors channel.
However, in this specific case, even that is not desired 😄 because the problem is not on the application side, it's just that the input message was not valid. There is nothing application could do about it and there is no reason for the application to crash because it did everything correctly. The best thing to do is just logging a regular message for visibility (why the input message was "ignored" and there was no output message) and move on.

TLDR replace `Fatalf` with `Printf` 😄

File: lecture_3/03_project/calculator/internal/infrastructure/sqlite/models/betCalculated.go [hunk 16]
```diff
@@ -1,7 +1,7 @@
 package models
 
-// BetCalculated is a storage model representation of a calculated bet.
-type BetCalculated struct {
+// Bet is a storage model representation of a bet.
+type Bet struct {
```
Comment (oreskdaxilis) [inline lecture_3/03_project/calculator/internal/infrastructure/sqlite/models/betCalculated.go]: File should now be renamed to `bet.go`

File: lecture_3/03_project/controller/internal/engine/engine.go (hunk 1 above)
Comment (oreskdaxilis) [reply lecture_3/03_project/controller/internal/engine/engine.go]: > I don't know if this is necessary, I think the bet should always be the same here


Yes if Calculator is the only service sending updates to Controller. Imagine we had another service, e.g. Payout service, whose purpose is to pay out money to customers and set `is_paid_out` field to `true` on the bet. Controller would pick up the change and send the latest state of that bet to bets queue, which will be consumed by Calculator. Ok, that specific field won't really be relevant for Calculator, but imagine it was - Calculator would need to update that bet in its internal database `calc_bets`.

**Comment by oreskdaxilis** [comment]
Also one more thing I have forgotten - don't commit databases, they don't belong to github.

File: lecture_3/03_project/calculator/internal/engine/handler/handler.go (hunk 15 above)
Comment (jelenanemcic) [reply lecture_3/03_project/calculator/internal/engine/handler/handler.go]: Thank you, I changed everything according to the suggestions. :)

**Review by oreskdaxilis — APPROVED** [review]
Well done 🎉
